### PR TITLE
New: Added HTML formatter (fixes #3505)

### LINKED ISSUE #3505: Built-in HTML formatter
ESLint doesn't seem to have a built-in HTML formatter, which seems like a very valid enhancement esp. when partnered with a task runner that possesses an "output to file" parameter (e.g `outputFile` in `grunt-eslint`). PR welcome?

### PR DESCRIPTION
**Pull request by JulianLaval**
Fixes #3505.

- Adds a built-in HTML formatter styled based on the current ESLint demo & documentation
- Automatically collapses/expands errors per file
- Handlebars added as dependency (@nzakas suggested Mustache, but it was a bit *too* logic-less for our needs), and Cheerio as devDependency for HTML unit tests.
- Added "html" to `-format` option documentation
- Full unit test coverage incl. edge cases such as illegal characters, missing properties

Demo JSFiddle: http://jsfiddle.net/julianlaval/sjzxkqm4/

### CODE REVIEW

**Comment by JulianLaval** [comment]
Fixes #3505

**Comment by JulianLaval** [comment]
Output is minified :)

**Comment by gyandeeps** [comment]
We need to have unit test.

**Comment by ilyavolodin** [comment]
@JulianLaval code looks good. As @gyandeeps mentioned, we need unit tests to accept any code PR. Also this [documentation](http://eslint.org/docs/user-guide/command-line-interface#f-format) should be updated as well.

**Comment by JulianLaval** [comment]
Ah yes, didn't check if formatters had unit tests. Will get on that and documentation.

**Comment by JulianLaval** [comment]
I've just noticed that the other formatters don't rely on the `errorCount` and `warningCount` properties for results, despite being documented. What is the preferred approach: using the existing properties or calculating them ourselves?

**Comment by gyandeeps** [comment]
I think almost all the formatters were created before we introduced those count properties that the reason they calculate themselves. But you should be good to use the properties. Thats their whole purpose basically.

**Comment by JulianLaval** [comment]
Perfect. Test cases will reflect this.

**Comment by JulianLaval** [comment]
PR updated!

File: lib/formatters/html.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    });
+}
+
+//------------------------------------------------------------------------------
+// Public Interface
+//------------------------------------------------------------------------------
+
+module.exports = function(results) {
+
+    var output = "<!DOCTYPE html><head><title>ESLint Report</title><style>"
```
Comment (nzakas) [inline lib/formatters/html.js]: Would it be easier to use Mustache to generate the HTML? This seems like a headache. :)

**Comment by JulianLaval** [comment]
@nzakas Aha, might indeed be better ;) Will have a go and update PR!

**Comment by JulianLaval** [comment]
Updated PR & description :)

File: lib/formatters/html.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    + "<table><tbody>{{#each results}}"
+    + "<tr class=\"bg-{{getColor this.errorCount this.warningCount}}\" data-group=\"f-{{@index}}\"><th colspan=\"4\">[+] {{this.filePath}} <span>{{renderText this.errorCount this.warningCount}}</span></th></tr>"
+    + "{{#each this.messages}}<tr class=\"f-{{@../index}}\" style=\"display:none\"><td>{{#if this.line}}{{this.line}}{{else}}0{{/if}}:{{#if this.column}}{{this.column}}{{else}}0{{/if}}</td>{{getSeverity this.severity}}<td>{{this.message}}</td><td><a href=\"http://eslint.org/docs/rules/{{this.ruleId}}\" target=\"_blank\">{{this.ruleId}}</a></td></tr>{{/each}}"
+    + "{{/each}}</tbody></table>"
+    + "<script type=\"text/javascript\">for(var groups=document.querySelectorAll(\"tr[data-group]\"),i=0;i<groups.length;i++)groups[i].addEventListener(\"click\",function(){var e=document.getElementsByClassName(this.getAttribute(\"data-group\"));this.innerHTML=this.innerHTML.indexOf(\"+\")>-1?this.innerHTML.replace(\"+\",\"-\"):this.innerHTML.replace(\"-\",\"+\");for(var t=0;t<e.length;t++)e[t].style.display=\"none\"!==e[t].style.display?\"none\":\"table-row\"});</script>"
+    + "</body></html>";
```
Comment (gyandeeps) [inline lib/formatters/html.js]: Minor but for readability purpose I would actually format this like 

```js
var template = "<!DOCTYPE html>" +
"<head>" +
    "<title>ESLint Report</title>" + 
    "<style>" +
        "body{font-family:Arial,\"Helvetica Neue\",Helvetica,sans-serif;font-size:16px;font-weight:normal;margin:0;padding:0;color:#333}#overview{padding:20px 30px}td,th{padding:5px 10px}h1{margin:0}table{margin:30px;min-width:50%;border-radius:5px;border:1px solid #ddd;border-spacing:0px;}th{font-weight:400;font-size:normal;text-align:left;cursor:pointer}td.clr-1,td.clr-2,th span{font-weight:700}th span{float:right;margin-left:20px}th span:after{content:\"\";clear:both;display:block}tr:last-child td{border-bottom:none}tr td:first-child,tr td:last-child{color:#9da0a4}#overview.bg-0,tr.bg-0 th{color:#468847;background:#dff0d8;border-bottom:1px solid #d6e9c6}#overview.bg-1,tr.bg-1 th{color:#f0ad4e;background:#fcf8e3;border-bottom:1px solid #fbeed5}#overview.bg-2,tr.bg-2 th{color:#b94a48;background:#f2dede;border-bottom:1px solid #eed3d7}td{border-bottom:1px solid #ddd}td.clr-1{color:#f0ad4e}td.clr-2{color:#b94a48}td a{color:#3a33d1;text-decoration:none}td a:hover{color:#272296;text-decoration:underline}" +
    "</style>" + 
"</head>" +
"<body>" +
    "<div id=\"overview\" class=\"bg-{{getColor totalErrors totalWarnings}}\">" +
        "<h1>ESLint Report</h1>" +
    "<div>" +
    "{{renderText totalErrors totalWarnings}} - Generated on {{date}}</div></div>" +
    "<table>" +
        "<tbody>" +
            "{{#each results}}" +
                "<tr class=\"bg-{{getColor this.errorCount this.warningCount}}\" data-group=\"f-{{@index}}\">" +
                    "<th colspan=\"4\">[+] {{this.filePath}} " +
                        "<span>{{renderText this.errorCount this.warningCount}}</span>" +
                    "</th>" +
                "</tr>" +
                "{{#each this.messages}}" + 
                    "<tr class=\"f-{{@../index}}\" style=\"display:none\">" +
                        "<td>{{#if this.line}}{{this.line}}{{else}}0{{/if}}:{{#if this.column}}{{this.column}}{{else}}0{{/if}}</td>" +
                        "{{getSeverity this.severity}}" +
                        "<td>{{this.message}}</td>" +
                        "<td>" +
                            "<a href=\"http://eslint.org/docs/rules/{{this.ruleId}}\" target=\"_blank\">{{this.ruleId}}</a>" +
                        "</td>" +
                    "</tr>" +
                "{{/each}}" +
            "{{/each}}" +
        "</tbody>" + 
    "</table>" +
    "<script type=\"text/javascript\">" +
        "for(var groups=document.querySelectorAll(\"tr[data-group]\"),i=0;i<groups.length;i++)groups[i].addEventListener(\"click\",function(){var e=document.getElementsByClassName(this.getAttribute(\"data-group\"));this.innerHTML=this.innerHTML.indexOf(\"+\")>-1?this.innerHTML.replace(\"+\",\"-\"):this.innerHTML.replace(\"-\",\"+\");for(var t=0;t<e.length;t++)e[t].style.display=\"none\"!==e[t].style.display?\"none\":\"table-row\"});" +
    "</script>" +
"</body>" +
"</html>";
```

Comment (JulianLaval) [inline lib/formatters/html.js]: Ha, much better, thanks. Admittedly a bit of laziness on my part; PR updated :)

**Comment by gyandeeps** [comment]
Thanks for the update but I think  you have accidentally copied your source file into your test file

**Comment by JulianLaval** [comment]
Oh fack, cheers for picking that up. Eek!

File: lib/formatters/html.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    return new Handlebars.SafeString((severity === 1) ? "<td class=\"clr-1\">Warning</td>" : "<td class=\"clr-2\">Error</td>");
+});
+
+//------------------------------------------------------------------------------
+// Public Interface
+//------------------------------------------------------------------------------
+
+module.exports = function(results) {
+
+    var template = "<!DOCTYPE html>" +
```
Comment (nzakas) [inline lib/formatters/html.js]: Can you move this into a separate handlebars file? It'll be easier to edit without all the string concatenation.

You can then read it in using fs.

File: tests/lib/formatters/html.js [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+                line: 5,
+                column: 10,
+                ruleId: "foo",
+                source: "foo"
+            }]
+        }];
+
+        it("should return a string in HTML format with 1 issue in 1 file and styled accordingly", function() {
+            var result = formatter(code);
+            assert.equal(parseHTML(result), "<div id=\"overview\" class=\"bg-2\"><h1>ESLint Report</h1><div>1 problem (1 error, 0 warnings)</div></div><table><tbody><tr class=\"bg-2\" data-group=\"f-0\"><th colspan=\"4\">[+] foo.js <span>1 problem (1 error, 0 warnings)</span></th></tr><tr class=\"f-0\" style=\"display:none\"><td>5:10</td><td class=\"clr-2\">Error</td><td>Unexpected foo.</td><td><a href=\"http://eslint.org/docs/rules/foo\" target=\"_blank\">foo</a></td></tr></tbody></table>");
```
Comment (nzakas) [inline tests/lib/formatters/html.js]: We probably don't want to be testing exact HTML string output as that will be hard to maintain. Maybe use Cheerio to look for specific things in the DOM?

File: lib/formatters/html.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    return (count === 1 ? word : word + "s");
+}
+
+/**
+ * Renders text along the template of x problems (x errors, x warnings)
+ * @param {string} totalErrors Total errors
+ * @param {string} totalWarnings Total warnings
+ * @returns {string} The formatted string, pluralized where necessary
+ */
+Handlebars.registerHelper("renderText", function(totalErrors, totalWarnings) {
```
Comment (nzakas) [inline lib/formatters/html.js]: We should probably create a new instance of handlebars so as not to pollute the global reference. 

`var handlebars = Handlebars.create()`

File: tests/lib/formatters/html.js (hunk 4 above)
Comment (JulianLaval) [inline tests/lib/formatters/html.js]: Noted! Too many extra dependencies isn't an issue?

File: lib/formatters/html.js (hunk 3 above)
Comment (JulianLaval) [inline lib/formatters/html.js]: Good shout. Any particular location, or simply `lib/formatters`?

File: lib/formatters/html.js (hunk 5 above)
Comment (JulianLaval) [inline lib/formatters/html.js]: Done deal :+1:

File: lib/formatters/html.js (hunk 3 above)
Comment (JulianLaval) [inline lib/formatters/html.js]: Likewise, for the sake of readability, should we unminify CSS & JS? Seems to add approx 60KB to the output

Comment (nzakas) [inline lib/formatters/html.js]: Yeah, there's no need to minify those.

I'd just stick the file in `lib/formatters` for now.

File: tests/lib/formatters/html.js (hunk 4 above)
Comment (nzakas) [inline tests/lib/formatters/html.js]: cheerio would be a dev dependency, so no issue there.

**Comment by JulianLaval** [comment]
PR updated! :+1:

**Comment by ilyavolodin** [comment]
LGTM
